.github/workflows/rust.yml

```diff
@@ -0,0 +1,46 @@
+name: Rust
+
+on:
+  push:
+    branches: [ "main" ]
+  pull_request:
+    branches: [ "main" ]
+
+env:
+  CARGO_TERM_COLOR: always
+
+jobs:
+  build:
+
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ubuntu-latest, macos-latest, windows-latest]
+
+
+    steps:
+    - uses: actions/checkout@v3
+      with:
+        submodules: 'true'
+    - name: Setup protoc
+      uses: arduino/setup-protoc@v1.1.2
+    - uses: actions-rs/toolchain@v1
+      with:
+        profile: minimal
+        toolchain: stable
+        override: true
+    - uses: Swatinem/rust-cache@v2
+    - name: Build
+      uses: actions-rs/cargo@v1
+      with:
+        command: build
+        args: --release
+    - name: Test
+      uses: actions-rs/cargo@v1
+      with:
+        command: test
+    - name: Upload artifacts 
+      uses: actions/upload-artifact@v2
+      with:
+        name: reveaal-${{ matrix.os }}
+        path: target/release/Reveaal
```

Comment:
Add formatting and clippy check (maybe just take the snippets from the previous)

Reply:
Whoops I rebased and lost those changes. I have added them back in.

Reply:
damn, inb4